test/UnitTests/MSTest.Analyzers.UnitTests/TypeContainingTestMethodShouldBeATestClassAnalyzer.cs

```diff
@@ -0,0 +1,161 @@
+
+            public class [|MyTestClass|]
+            {
+                [TestMethod]
+                public void TestMethod1() {}
+            }
+            """;
+
+        await VerifyCS.VerifyAnalyzerAsync(code);
+    }
+
+    public async Task WhenClassWithoutTestAttribute_AndWithoutTestMethods_InheritTestClassWithTestMethods_Diagnostic()
+    {
+        string code = """
+            using Microsoft.VisualStudio.TestTools.UnitTesting;
+
+            public class [|MyTestClass|] : WithTestMethods_WithoutTestClass
+            {
+
+            }
+
+            [TestClass]
+            public class WithTestMethods_WithoutTestClass
+            {
+                [TestMethod]
+                public void TestMethod1()
+                {
+                }
+
+                [TestMethod]
+                public void TestMethod2()
+                {
+                }
+            }
+            """;
+
+        await VerifyCS.VerifyAnalyzerAsync(code);
+    }
+
+    public async Task WhenClassWithoutTestAttribute_AndWithTestMethods_InheritTestClass_Diagnostic()
+    {
+        string code = """
+            using Microsoft.VisualStudio.TestTools.UnitTesting;
+
+            [TestClass]
+            public class Base
+            {
+
+            }
+
+            public class [|MyTestClass|] : Base
+            {
+                [TestMethod]
+                public void TestMethod1()
+                {
+                }
+
+                [TestMethod]
+                public void TestMethod2()
+                {
+                }
+            }
+            """;
+
+        await VerifyCS.VerifyAnalyzerAsync(code);
+    }
+
+    public async Task WhenInheritedTestClassAttribute_HasInheritedTestMethodAttribute_NoDiagnostic()
+    {
+        string code = """
+            using Microsoft.VisualStudio.TestTools.UnitTesting;
+
+            public class DerivedTestMethod : TestMethodAttribute
+            {
+            }
+
+            public class DerivedTestClass : TestClassAttribute
+            {
+            }
+
+            [DerivedTestClass]
+            public class MyTestClass
+            {
+                [DerivedTestMethod]
+                public void MyTestMethod()
+                {
+                }
+            }
+            """;
+
+        await VerifyCS.VerifyAnalyzerAsync(code);
+    }
+
+    public async Task WhenClassWithoutTestAttribute_HasInheritedTestMethodAttribute_Diagnostic()
+    {
+        string code = """
+            using Microsoft.VisualStudio.TestTools.UnitTesting;
+
+            public class DerivedTestMethod : TestMethodAttribute
+            {
+            }
+
+            public class [|MyTestClass|]
+            {
+                [DerivedTestMethod]
+                public void MyTestMethod()
+                {
+                }
+            }
+            """;
+
+        await VerifyCS.VerifyAnalyzerAsync(code);
+    }
+
+    public async Task WhenAbstractClass_DoesNotHaveTestMethod_NoDiagnostic()
```

Comment:
I think you should test that there is no warning when the class **has** test method and is abstract but not marked with `[TestClass]`.